crates/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -35,8 +35,13 @@ impl Ssa {
     /// changes. This is because if the function's id later becomes known by a later
     /// pass, we would need to re-run all of inlining anyway to inline it, so we might
     /// as well save the work for later instead of performing it twice.
-    pub(crate) fn inline_functions(self) -> Ssa {
-        InlineContext::new(&self).inline_all(self)
+    pub(crate) fn inline_functions(mut self) -> Ssa {
```

Comment:
I think we should consider not treating every brillig function as an entry point, since inlining brillig functions will be beneficial for runtime cost and for applying further optimizations. We could treat as entry points all brillig functions called from ACIR and all recursive functions. Collecting those should be easy. Then in the inline code itself, when encountering a call to brillig, do something like this
```rust
 RuntimeType::Brillig => {
                            // Do not inline brillig into acir and don't inline recursion.
                            if self.source_function.runtime() == RuntimeType::Brillig
                                && !self.context.entry_point_functions.contains(&function)
                            {
                                self.inline_function(ssa, *id, function, arguments);
                            } else {
                                self.push_instruction(*id);
                            }
                        }
```

What do you think?

Reply:
Instead of the entry point functions check, I think a check to prohibit inlining brillig into Acir would be more straightforward. I can add it, although this PR is on hold currently until I break out the required mem2reg updates.

Reply:
I'm holding off on this optimizing brillig into brillig optimization since we'd need a check to fail if we recur too many times and this requires more code. I think we can add it in a later PR.

Reply:
Makes sense!